### PR TITLE
Add fxa/sync integration tests

### PR DESCRIPTION
**Pull request by isabelrios**
Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [-] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [-] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

This PR tries to add new tests, sync integration tests, to check the sync process Desktop<->Fenix, first for Bookmarks and in the future for more.

### CODE REVIEW

**Comment by isabelrios** [comment]
@rpappalax, @npark-mozilla once the checks are green I would like to ask you to take a first look at the changes in this PR about adding the sync integration tests.
In case of questions please let me know or check the [doc with all the implementation details](https://docs.google.com/document/d/1dhxlbGQBA6aJi2Xz-CsJZuGJPRReoL7nfm9cYu4HcZI/edit?usp=sharing) for more info...
Thanks!!

File: app/src/androidTest/resources/email.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+pytest-28f90ad8b4@restmail.net
```
Comment (isabelrios) [inline app/src/androidTest/resources/email.txt]: Note that this is not a permanent account, while running the tests via `pipenv run pytest`a new account will be created, written in this file and read after that for the test to be used to sync in Fenix app. 
Same for the password below.

File: automation/taskcluster/androidTest/flank-x86.yml [hunk 2]
```diff
@@ -33,6 +33,9 @@ gcloud:
     - /sdcard/screenshots
   performance-metrics: true
 
+  test-targets:
```
Comment (isabelrios) [inline automation/taskcluster/androidTest/flank-x86.yml]: Adding this line and below we avoid running these sync integration tests as part as the regular UI tests.

File: app/src/androidTest/java/org/mozilla/fenix/Experiments.kt [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix
+
+import android.content.Context
+import mozilla.components.service.fretboard.ExperimentDescriptor
+
+const val EXPERIMENTS_JSON_FILENAME = "experiments.json"
+const val EXPERIMENTS_BASE_URL = "https://firefox.settings.services.mozilla.com/v1"
```
Comment (rpappalax) [inline app/src/androidTest/java/org/mozilla/fenix/Experiments.kt]: this looks like a production server.  Is that intentional for this test?

File: app/src/androidTest/java/org/mozilla/fenix/components/FxaServer.kt [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package org.mozilla.fenix.components
+
+import android.content.Context
+import mozilla.components.service.fxa.ServerConfig
+
+object FxaServer {
+    const val CLIENT_ID = "a2270f727f45f648"
```
Comment (rpappalax) [inline app/src/androidTest/java/org/mozilla/fenix/components/FxaServer.kt]: will this CLIENT_ID also be replaced by a dynamic one?  (same as:  https://github.com/mozilla-mobile/fenix/pull/5633/files/b6603729b13964d8f52aa06342a74601f2bf6cd3#r329536741)?  It's OK to hard code these if necessary, but it would be more brittle.  Esp. on the stage server

File: app/src/androidTest/java/org/mozilla/fenix/syncintegration/Pipfile [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+mozdownload = "*"
+mozinstall = "*"
+mozprofile = "*"
+mozrunner = "*"
+mozversion = "*"
+pytest = "*"
+pytest-fxa = "*"
+pytest-html = "*"
+pytest-metadata = "*"
+requests = "*"
```
Comment (rpappalax) [inline app/src/androidTest/java/org/mozilla/fenix/syncintegration/Pipfile]: do you not want to pin down these version numbers for stability?

File: app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    /* These tests will be running in the future
+    // once the test above runs successfully and
+    // the environment is stable
+
+    // Bookmark item Desktop -> Fenix
+    @Test
+    fun checkBookmarkFromDesktopTest() {
+        signInFxSync()
+        tapReturnToPreviousApp()
+        sleep(5000)
```
Comment (rpappalax) [inline app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt]: i imagine you're going to refine this later, but I recommend creating a constant for this in the helpers.  We currently have a few wait times defined in https://github.com/mozilla-mobile/fenix/blob/master/app/src/androidTest/java/org/mozilla/fenix/helpers/TestAssetHelper.kt which was created for Mock server assets / functions.  With my next PR, I'll be adding a more generic https://github.com/mozilla-mobile/fenix/blob/master/app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt which would be for misc functions and constants.  We could eventually migrate our constants there.

File: app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        mDevice.pressBack()
+        signInFxSync()
+    }
+
+    // Bookmark item Fenix -> Desktop
+    @Test
+    fun checkHistoryFromDeviceTest() {
+        tapInToolBar()
+        typeInToolBar()
+        sleep(3000)
```
Comment (rpappalax) [inline app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt]: see above comment

File: app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        signInFxSync()
+    }
+    */
+
+    // Useful functions for the tests
+    fun typeEmail() {
+        val emailInput = mDevice.findObject(UiSelector()
+                .instance(0)
+                .className(EditText::class.java))
+        emailInput.waitForExists(1000)
```
Comment (rpappalax) [inline app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt]: see above

File: app/src/androidTest/java/org/mozilla/fenix/components/FxaServer.kt (hunk 4 above)
Comment (isabelrios) [reply app/src/androidTest/java/org/mozilla/fenix/components/FxaServer.kt]: This implementation is similar, using same CLIENT_ID for the non test part, the regular FxaServer.kt. But I can check if this can be changed

File: app/src/androidTest/java/org/mozilla/fenix/syncintegration/Pipfile (hunk 5 above)
Comment (isabelrios) [reply app/src/androidTest/java/org/mozilla/fenix/syncintegration/Pipfile]: We have had this for the iOS tests for long time and did not face any stability issues...but if you prefer I change that, I can look into that.

File: app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt (hunk 8 above)
Comment (isabelrios) [reply app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt]: I will change that to use the constant in helpers. Sorry I forgot about that.

File: app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    val activityTestRule = HomeActivityTestRule()
+
+    // History item Desktop -> Fenix
+    @Test
+    fun checkHistoryFromDesktopTest() {
+        signInFxSync()
+        tapReturnToPreviousApp()
+        homeScreen {
+        }.openThreeDotMenu {}
+        libraryButton()
```
Comment (npark-mozilla) [inline app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt]: this is just a style preference, but I put these methods inside the bracket of `openThreeDotMenu` to show that these actions are happening within the three dot menu dialog.  For me it's more easier to read.

File: app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        passwordInput.setText(passwordValue)
+    }
+
+    fun tapOnSygIn() {
+        mDevice.pressEnter()
+        mDevice.wait(Until.findObjects(By.text("Sign in")), 3000)
+        val signInButton = mDevice.findObject(UiSelector()
+                .instance(0)
+                .className(Button::class.java))
+        signInButton.waitForExists(10000)
```
Comment (npark-mozilla) [inline app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt]: use constants

File: app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+                .className(EditText::class.java))
+        emailInput.waitForExists(1000)
+
+        val emailAddress = javaClass.classLoader.getResource("email.txt").readText()
+        emailInput.setText(emailAddress)
+    }
+
+    fun tapOnContinueButton() {
+        val continueButton = mDevice.findObject(By.res("submit-btn"))
+        continueButton.clickAndWait(Until.newWindow(), 50000)
```
Comment (npark-mozilla) [inline app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt]: use constants

File: app/src/main/java/org/mozilla/fenix/components/FxaServer.kt [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+import org.mozilla.fenix.isInExperiment
+
+object FxaServer {
+    const val CLIENT_ID = "a2270f727f45f648"
+    const val REDIRECT_URL = "https://accounts.firefox.com/oauth/success/$CLIENT_ID"
+
+    fun redirectUrl(context: Context) = if (context.isInExperiment(Experiments.asFeatureWebChannelsDisabled)) {
+        REDIRECT_URL
+    } else {
+        "urn:ietf:wg:oauth:2.0:oob:oauth-redirect-webchannel"
```
Comment (npark-mozilla) [inline app/src/main/java/org/mozilla/fenix/components/FxaServer.kt]: should we put this as a constant?

File: app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt (hunk 9 above)
Comment (isabelrios) [reply app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt]: Oh right, when I created this PR those options were not added to the robot menu yet, so I added functions at the end of the test. I will replace them with the actions from robot!

**Comment by isabelrios** [comment]
@csadilek asking for your review since you are familiar with all the changes at least for a first deep look..thanks!

File: app/src/main/java/org/mozilla/fenix/components/FxaServer.kt [hunk 13]
```diff
@@ -0,0 +1,21 @@
+package org.mozilla.fenix.components
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/components/FxaServer.kt]: We need a license header here :).

File: app/src/main/java/org/mozilla/fenix/components/FxaServer.kt [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+
+import android.content.Context
+import mozilla.components.service.fxa.ServerConfig
+import org.mozilla.fenix.Experiments
+import org.mozilla.fenix.isInExperiment
+
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/components/FxaServer.kt]: Let's add some KDocs here e.g.

```
/**
 * Utility to configure Firefox Account servers.
 */
```

File: docs/syncIntegrationTests.md [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+The process for example for History item Desktop -> Fenix, would be:
+- Desktop is launched, user signed in and history item created.
+- Android sim is launched (Pixel 3 API28), Fenix app starts and same user is signed in, then we go to History list and verify that the item is there.
+
+
+### Results
+Due to the set up necessary these tests do not run as part of the regular CI, via Taskcluster. 
+The idea is to have them running on Jenkins periodically (TBD how often).
+Once they finish there is a slack notificattion received informing about the result (so far that is configured for #firefox-ios-alerts)
```
Comment (csadilek) [inline docs/syncIntegrationTests.md]: ios? :)

File: docs/syncIntegrationTests.md [hunk 16]
```diff
@@ -0,0 +1,26 @@
+### Sync Integration Tests
```
Comment (csadilek) [inline docs/syncIntegrationTests.md]: Thanks for writing these docs!

File: app/src/androidTest/java/org/mozilla/fenix/components/FxaServer.kt [hunk 17]
```diff
@@ -0,0 +1,17 @@
+package org.mozilla.fenix.components
```
Comment (csadilek) [inline app/src/androidTest/java/org/mozilla/fenix/components/FxaServer.kt]: We forgot the license header.

File: app/src/androidTest/java/org/mozilla/fenix/components/FxaServer.kt [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+
+import android.content.Context
+import mozilla.components.service.fxa.ServerConfig
+
+object FxaServer {
```
Comment (csadilek) [inline app/src/androidTest/java/org/mozilla/fenix/components/FxaServer.kt]: Let's add some KDocs, see my comment below.

File: app/src/androidTest/java/org/mozilla/fenix/Experiments.kt [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+    // application services flag to disable Firefox Accounts WebChannel integration.
+    val asFeatureWebChannelsDisabled = ExperimentDescriptor("asFeatureWebChannelsDisabled")
+}
+
+val Context.app: FenixApplication
+    get() = applicationContext as FenixApplication
+
+fun Context.isInExperiment(descriptor: ExperimentDescriptor): Boolean =
+    if (descriptor.name == "asFeatureWebChannelsDisabled")
+        true
```
Comment (grigoryk) [inline app/src/androidTest/java/org/mozilla/fenix/Experiments.kt]: The only reason webchannel stuff didn't work is because we didn't whitelist stage FxA server in the webchannel extension manifest file. We certainly do want these tests to exercise webchannel flow, since that's the flow we ship by default. This can be an easy follow-up PR though.

File: app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+    fun typePassowrd() {
+        val passwordInput = mDevice.findObject(UiSelector()
+                .instance(0)
+                .className(EditText::class.java))
+
+        val passwordValue = javaClass.classLoader.getResource("password.txt").readText()
+        passwordInput.setText(passwordValue)
+    }
+
+    fun tapOnSygIn() {
```
Comment (grigoryk) [inline app/src/androidTest/java/org/mozilla/fenix/syncintegration/SyncIntegrationTest.kt]: nit: typo?

File: app/src/test/java/org/mozilla/fenix/components/BackgroundServicesTest.kt [hunk 21]
```diff
@@ -50,10 +50,10 @@ class BackgroundServicesTest {
         val context = mockk<Context>(relaxed = true)
 
         every { context.isInExperiment(eq(Experiments.asFeatureWebChannelsDisabled)) } returns false
-        assertEquals("urn:ietf:wg:oauth:2.0:oob:oauth-redirect-webchannel", BackgroundServices.redirectUrl(context))
+        // assertEquals("urn:ietf:wg:oauth:2.0:oob:oauth-redirect-webchannel", BackgroundServices.redirectUrl(context))
```
Comment (grigoryk) [inline app/src/test/java/org/mozilla/fenix/components/BackgroundServicesTest.kt]: This seems wrong. Why did you need to disable these tests?

Comment (csadilek) [reply app/src/test/java/org/mozilla/fenix/components/BackgroundServicesTest.kt]: Ah this was likely because `redirectUrl` is no longer a function on `BackgroundServices`. Didn't notice this was commented out. So, we can change that to `FxaServer.redirectUrl(context)` now and it should work again.

Comment (isabelrios) [reply app/src/test/java/org/mozilla/fenix/components/BackgroundServicesTest.kt]: Oh, sorry about that I commented it so that building on TC did not complain and forgot to fix. Fixing...

**Comment by isabelrios** [comment]
> This is really neat. How do you ensure that FxA signin verification won't be necessary during a test?

The verification happens after the account is created with [fxacli](https://pypi.org/project/fxacli/) tool in `plugin.py` (part of the code):

```
....
fxa_account = FxAccount(email=account.email, password=password)
    try:
        session = fxa_client.create_account(fxa_account.email,
                                            fxa_account.password)
        logger.info('Created: {}'.format(fxa_account))
        account.fetch()
        message = account.wait_for_email(
            lambda m: 'x-verify-code' in m['headers'] and
            session.uid == m['headers']['x-uid']
        )
        assert (message is not None), 'Verification email not found.'
        session.verify_email_code(message['headers']['x-verify-code'])
        logger.info('Verified: {}'.format(fxa_account))
        yield fxa_account
....
```

In the logs you can see something as:
```
root:plugin.py:93 Created: FxAccount(email='pytest-9e83976acb@restmail.net', password='cwyOfVGK')
root:plugin.py:101 Verified: FxAccount(email='pytest-9e83976acb@restmail.net', password='cwyOfVGK')
```

File: docs/syncIntegrationTests.md (hunk 15 above)
Comment (isabelrios) [reply docs/syncIntegrationTests.md]: Oh we set up that channel for looking at the results of the similar tests that we have on iOS, instead of creating a new channel I'm checking the results in that one, but I could create a new one and report there...

Comment (csadilek) [reply docs/syncIntegrationTests.md]: Ah no that's fine, whatever works for you. Just wanted to make sure it's not a typo, doesn't need to block landing.

**Comment by sblatz** [comment]
Is this ready to land? This PR has been up and approved for a while. Want to make sure I don't merge it before all nits have been addressed though.

**Comment by isabelrios** [comment]
@sblatz yes, it should be ready, but if you want me to double check with one (or more) reviewers, let me know...

**Comment by isabelrios** [comment]
> > This is really neat. How do you ensure that FxA signin verification won't be necessary during a test?
> 
> The verification happens after the account is created with [fxacli](https://pypi.org/project/fxacli/) tool in `plugin.py` (part of the code):
> 
> ```
> ....
> fxa_account = FxAccount(email=account.email, password=password)
>     try:
>         session = fxa_client.create_account(fxa_account.email,
>                                             fxa_account.password)
>         logger.info('Created: {}'.format(fxa_account))
>         account.fetch()
>         message = account.wait_for_email(
>             lambda m: 'x-verify-code' in m['headers'] and
>             session.uid == m['headers']['x-uid']
>         )
>         assert (message is not None), 'Verification email not found.'
>         session.verify_email_code(message['headers']['x-verify-code'])
>         logger.info('Verified: {}'.format(fxa_account))
>         yield fxa_account
> ....
> ```
> 
> In the logs you can see something as:
> 
> ```
> root:plugin.py:93 Created: FxAccount(email='pytest-9e83976acb@restmail.net', password='cwyOfVGK')
> root:plugin.py:101 Verified: FxAccount(email='pytest-9e83976acb@restmail.net', password='cwyOfVGK')
> ```
> 
> When the test finishes, the account is removed.

@grigoryk I replied to your question but did not mention you so not sure if you read it and that answers your question...Thanks!

**Comment by isabelrios** [comment]
Hi @sblatz , what's that bors expected? Should I do something before merging?

**Comment by sblatz** [comment]
bors r=csadilek
